hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/AbstractGrantApprovalCall.java

```diff
@@ -60,32 +64,89 @@ protected AbstractGrantApprovalCall(
+            return buildCryptoApproveAllowance(approve(senderId));
+        }
+    }
+
+    private CryptoDeleteAllowanceTransactionBody remove(AccountID ownerId) {
+        return CryptoDeleteAllowanceTransactionBody.newBuilder()
+                .nftAllowances(NftRemoveAllowance.newBuilder()
+                        .tokenId(token)
+                        .owner(ownerId)
+                        .serialNumbers(amount.longValue())
+                        .build())
                 .build();
     }
 
-    private CryptoApproveAllowanceTransactionBody approve(
-            @NonNull final TokenID token,
-            @NonNull final AccountID spender,
-            @NonNull final BigInteger amount,
-            @NonNull final TokenType tokenType) {
+    private CryptoApproveAllowanceTransactionBody approveDelegate(AccountID ownerId, AccountID delegateSpenderId) {
+        return CryptoApproveAllowanceTransactionBody.newBuilder()
+                .nftAllowances(NftAllowance.newBuilder()
+                        .tokenId(token)
+                        .spender(spender)
+                        .delegatingSpender(delegateSpenderId)
+                        .owner(ownerId)
+                        .serialNumbers(amount.longValue())
+                        .build())
+                .build();
+    }
+
+    private CryptoApproveAllowanceTransactionBody approve(AccountID ownerId) {
         return tokenType.equals(TokenType.FUNGIBLE_COMMON)
                 ? CryptoApproveAllowanceTransactionBody.newBuilder()
                         .tokenAllowances(TokenAllowance.newBuilder()
                                 .tokenId(token)
                                 .spender(spender)
-                                .owner(senderId)
+                                .owner(ownerId)
                                 .amount(amount.longValue())
                                 .build())
                         .build()
                 : CryptoApproveAllowanceTransactionBody.newBuilder()
                         .nftAllowances(NftAllowance.newBuilder()
                                 .tokenId(token)
                                 .spender(spender)
-                                .owner(senderId)
+                                .owner(ownerId)
                                 .serialNumbers(amount.longValue())
                                 .build())
                         .build();
     }
+
+    private TransactionBody buildCryptoDeleteAllowance(CryptoDeleteAllowanceTransactionBody body) {
+        return TransactionBody.newBuilder().cryptoDeleteAllowance(body).build();
+    }
+
+    private TransactionBody buildCryptoApproveAllowance(CryptoApproveAllowanceTransactionBody body) {
+        return TransactionBody.newBuilder().cryptoApproveAllowance(body).build();
+    }
+
+    private AccountID getOwnerId() {
+        final var nft = enhancement.nativeOperations().getNft(token.tokenNum(), amount.longValue());
+        return nft != null && nft.hasOwnerId()
```

Comment:
If the NFT was missing, this would return its token's treasury as its owner `AccountID`.

At this point have we verified the NFT exists? In that case we could use `requireNonNull(nft)` here?

Reply:
Noted. Thanks!